app/upgrades/testnet/v4/upgrades.go

```diff
@@ -0,0 +1,64 @@
+package tv4
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	slashingkeeper "github.com/cosmos/cosmos-sdk/x/slashing/keeper"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+
+	feemarketkeeper "github.com/tharsis/ethermint/x/feemarket/keeper"
+	feemarkettypes "github.com/tharsis/ethermint/x/feemarket/types"
+
+	claimskeeper "github.com/tharsis/evmos/v3/x/claims/keeper"
+)
+
+// CreateUpgradeHandler creates an SDK upgrade handler for v4. It handles the upgrade
+// functionality by importing the exported state. It doesn't perform any in-place store
+// migrations for the modules since there are no consensus version bumps. Aditionally,
+// it updates the slashing, fee market and claims params.
+
+func CreateUpgradeHandler(
```

Comment:
why do we need a testnet and mainnet upgradeHandler? shouldn't they always be the same?

Reply:
Ideally yes. I think it sucks that our testnet doesn't act like a staging environment for mainnet and therefore we're never actually testing the mainnet upgrade handler.

My suggestion would be to teardown testnet at some point, delete the upgrade handlers, then obtain parity with the mainnet upgrades going forward, treating it like a staging environment. Alternatively, have multiple testnets (bad for users who want to test on there).

Reply:
Alternatively, we create a testing CI for specifically for testing upgrades and then just forget testnet manual upgrades.

Reply:
> why do we need a testnet and mainnet upgradeHandler? shouldn't they always be the same?

Yes, they should ideally be the same, but the upgrade heights and chain-id are different, so there needs to be specific logic to handle upgrades depending on those values

Reply:
updated